### DISCUSSION TITLE
关于除 Markdown 外其他标记语言支持的可能

### DISCUSSION THREAD

**Discussion by pure01fx** [YunYouJun/valaxy#432] [General]
_原谅我对前端的一窍不通所以没有认真浏览过代码，_ 在 Hexo 中有 Renderer 这种 API，我现在是通过其解析 Typst 的文档（利用 Pandoc 处理 Hexo 给出的文档内容（去掉了 front-matter 之后的内容） + 一些对于 Pandoc 生成的 DOM 的后处理（例如给子标题加 id 方便目录跳转，等等），不过仍然保留 Nunjucks 标签然后再给 Hexo 进行后处理）

不知道作者有没有考虑过对于其他标记语言的支持，或者如果有这方面的设计的话是否有这方面的 API，因为我个人认为博客的大部分功能不一定是 Markdown-exclusive 的，只要是能够生成 HTML 内容的语言支持起来应该差不多。

以 Markdown 为首要支持目标是正确的，不过我觉得如果可以的话能够抽象个 Renderer 也挺好的，虽然支持的功能少点但是这部分支持的内容可以靠 Renderer 的 Dev 去手搓 HTML 来自己弥补一些。 _当然也不排除现在已经有这一层抽象了，如果是这样的话很抱歉，我没有发现（_

补充个示例：
![image](https://github.com/user-attachments/assets/bef99b4b-34ad-4613-8928-688a74ef78f9)
![image](https://github.com/user-attachments/assets/adfaefa3-7aa7-43b4-89a5-64b2b934adf2)

简化的 Renderer：
```js
function pandocRenderer(data, options) {
    let pandoc_path = hexo.config?.pandoc?.pandoc_path ?? 'pandoc';
    let args = parseArgs.bind(hexo)(data, options);
    let src = data.text.toString();

    const cwd = get_directory(data.path) ?? process.cwd();

    let res = spawnSync(pandoc_path, args, {
        cwd: cwd,
        env: process.env,
        encoding: "utf8",
        input: src.replace(/.+?\/\/ HEXO: /g, '')
    });

    if (res.status === 0) {
        if (res.stderr) { /* ... */ }

        return processResult(res.stdout);
    } else { /* ... */ }
}

hexo.extend.renderer.register('typ', 'html', pandocRenderer, true);
```

**Comment by YunYouJun** [comment]
其他标记语言是指什么？
渲染器基于 markdown-it 实现，可以通过 valaxy.config.ts `markdown` 为  markdown-it 使用插件。

**Reply by doudou0720** [reply]
我觉得可能有 `.adoc` (AsciiDoc)? @YunYouJun 

我记得Github好像还支持渲染这个文档，比如这个仓库：https://github.com/asciidoctor/asciidoctor

**Comment by YunYouJun** [comment]
感谢 @pure01fx 和 @doudou0720 的建议 🙏 统一回复下「支持 Markdown 以外标记语言」这件事。

**先说结论:对博客场景，这个需求其实意义不大，因为想要的东西基于 Markdown 基本都能实现。**

正如 @pure01fx 自己提到的，博客的大部分功能并不是某种标记语言独有的。在 Valaxy 里：

- markdown-it 可以通过 `valaxy.config.ts` 的 `markdown` 字段自由挂插件（自定义容器、语法等）；
- 数学公式已内置（KaTeX / MathJax），脚注、目录、`:::` 容器、`include` 片段等也都有；
- 最关键的一点：Valaxy 的 Markdown 会编译成 **Vue SFC**，所以任何 AsciiDoc / Typst 能产出的 HTML，你都可以在 `.md` 里直接写组件 / 原生 HTML 来实现，甚至能进一步做交互。

也就是说，换一门标记语言能额外带来的东西很有限，成本却不小。

**为什么不打算做进核心**：Valaxy 不像 Hexo 那样把文档渲染成 HTML 字符串再塞进模板，而是渲染成 Vue SFC，所以 Hexo 那套 `renderer.register('typ', 'html', fn)`「给我 HTML」的契约没法照搬——一个 renderer 在 Valaxy 里实际要产出 Vue SFC，并自己接上 frontmatter、excerpt、目录（headers）、加密、本地搜索索引、代码高亮一整条管线；否则很容易变成「`.adoc` 没目录 / 搜不到」这类问题。此外 Pandoc / Typst 方案依赖外部二进制，对 CI、Netlify/Vercel 等部署都不友好；Typst 的 HTML 导出目前也还在实验阶段（成熟的是 →PDF）。

我也暂不打算专门为此抽一套公开 API——一个公开扩展点本身就是长期维护成本，在没有真实使用者时也很难一次设计对。

不过**门是开着的**：如果确实有人想用（尤其 AsciiDoc——Asciidoctor.js 是纯 JS、无需外部二进制、自带目录，阻力最小），欢迎先做一个原型（独立 Vite 插件 / addon 都行）。只要有真实可跑的实现来验证契约，我很乐意在核心开一个小口子（把文件扩展名、`include` 规则等做成可配置），让它以 addon 形式接入。

**Comment by doudou0720** [comment]
很抱歉，但是我本人现在仍然在读高二，看看高考之后有没有时间了




艺艺
***@***.***
